huaweicloud/services/cfw/resource_huaweicloud_cfw_dns_resolution.go

```diff
@@ -0,0 +1,395 @@
+	}
+
+	userDefaultDNSServers := d.Get("default_dns_servers").(*schema.Set).List()
+	constains := utils.StrSliceContainsAnother(utils.ExpandToStringList(defaultDNSServers), utils.ExpandToStringList(userDefaultDNSServers))
+	if !constains {
+		return diag.Errorf("error creating CFW DNS resolution configuration: some non-default DNS servers have been specified")
+	}
+
+	userCustomServers := d.Get("custom_dns_servers").(*schema.Set).List()
+	if len(userCustomServers) == 0 && len(userDefaultDNSServers) == 0 {
+		return diag.Errorf("error creating CFW DNS resolution configuration: no DNS servers have been specified")
+	}
+
+	path := client.Endpoint + httpUrl
+	path = strings.ReplaceAll(path, "{project_id}", client.ProjectID)
+	path += fmt.Sprintf("?fw_instance_id=%s", fwInstanceID)
+
+	opt := golangsdk.RequestOpts{
+		KeepResponseBody: true,
+	}
+	opt.JSONBody = utils.RemoveNil(buildDNSResolutionConfigurationBodyParams(d, defaultDNSServers))
+	resp, err := client.Request("PUT", path, &opt)
+	if err != nil {
+		return diag.Errorf("error creating CFW DNS resolution configuration: %s", err)
+	}
+
+	_, err = utils.FlattenResponse(resp)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	d.SetId(fwInstanceID)
+
+	return resourceDNSResolutionRead(ctx, d, meta)
+}
+
+func readDNSServers(client *golangsdk.ServiceClient, id string) (interface{}, error) {
+	httpUrl := "v1/{project_id}/dns/servers"
+	path := client.Endpoint + httpUrl
+	path = strings.ReplaceAll(path, "{project_id}", client.ProjectID)
+	path += "?offset=0&limit=100"
+	path += fmt.Sprintf("&fw_instance_id=%s", id)
+
+	resp, err := pagination.ListAllItems(
+		client,
+		"offset",
+		path,
+		&pagination.QueryOpts{MarkerField: ""})
+
+	if err != nil {
+		return nil, fmt.Errorf("error retrieving CFW DNS resolution configuration: %s", err)
+	}
+
+	respJson, err := json.Marshal(resp)
+	if err != nil {
+		return nil, fmt.Errorf("error marshaling CFW DNS resolution configuration: %s", err)
+	}
+
+	var respBody interface{}
+	err = json.Unmarshal(respJson, &respBody)
+	if err != nil {
+		return nil, fmt.Errorf("error unmarshaling CFW DNS resolution configuration: %s", err)
+	}
+
+	return respBody, nil
+}
+
+func buildDNSResolutionConfigurationBodyParams(d *schema.ResourceData, defaultDNSServers []interface{}) map[string]interface{} {
+	userDefaultServers := d.Get("default_dns_servers").(*schema.Set).List()
+	userCustomServers := d.Get("custom_dns_servers").(*schema.Set).List()
+	differenceServers := difference(userDefaultServers, defaultDNSServers)
+
+	return map[string]interface{}{
+		"dns_server":               buildDNSServersBodyParams(userDefaultServers, differenceServers, userCustomServers),
+		"health_check_domain_name": utils.ValueIngoreEmpty(d.Get("health_check_domain_name")),
+	}
+}
+
+func difference(userDefaultServers, defaultDNSServers []interface{}) []interface{} {
+	userDefaultServersSet := make(map[string]bool)
```

Comment:
```suggestion
	userDefaultServersSet := make(map[string]struct{})
```
change it to struct is more effective

Reply:
fixed